src/CP15.cpp

```diff
@@ -332,153 +366,596 @@ u32 ARMv5::RandomLineIndex()
+    if (CP15BISTTestStateRegister & CP15_BIST_TR_DISABLE_DCACHE_LINEFILL)
+    {
+        DataCycles = NDS.ARM9MemTimings[tag >> 14][2]; 
+        if (addr < ITCMSize)
+        {
+            return *(u32*)&ITCM[addr & (ITCMPhysicalSize - 3)];
+        } else
+        if ((addr & DTCMMask) == DTCMBase)
+        {
+            return *(u32*)&DTCM[addr & (DTCMPhysicalSize - 3)];
+        } else
+        {
+            return BusRead32(addr & ~3);
+        }        
     }
-    if (ICacheTags[id+1] == tag)
+
+    u32 line;
+#if 0
+    // caclulate in which cacheline the data is to be filled
+    // The code below is doing the same as the if-less below
+    // It increases performance by reducing banches. 
+    // The code is kept here for readability.
+    // 
+    // NOTE: If you need to update either part, you need
+    //       to update the other too to keep them in sync!
+    //  
+
+    if (CP15Control & CP15_CACHE_CR_ROUNDROBIN)
     {
-        ICacheTags[id+1] = 1;
-        return;
+        line = DCacheCount;
+        DCacheCount = (line+1) & (DCACHE_SETS-1);
     }
-    if (ICacheTags[id+2] == tag)
+    else
     {
-        ICacheTags[id+2] = 1;
-        return;
+        line = DCacheRandom();
     }
-    if (ICacheTags[id+3] == tag)
+
+    // Update the selected set depending on the DCache LockDown register
+    if (DCacheLockDown)
     {
-        ICacheTags[id+3] = 1;
-        return;
+        if (DCacheLockDown & CACHE_LOCKUP_L)
+        {
+            // load into locked up cache
+            // into the selected set
+            line = (DCacheLockDown & (DCACHE_SETS-1)) + id;
+        } else
+        {
+            u8 minSet = ICacheLockDown & (DCACHE_SETS-1);
+            line = (line | minSet) + id;
+        }
+    }
+#else
+    // Do the same as above but instead of using if-else 
+    // utilize the && and || operators to skip parts of the operations
+    // With the order of comparison we can put the most likely path
+    // checked first
+
+    bool doLockDown = (DCacheLockDown & CACHE_LOCKUP_L);
+    bool roundRobin = CP15Control & CP15_CACHE_CR_ROUNDROBIN;
+    (!roundRobin && (line = RandomLineIndex())) || (roundRobin && (DCacheCount = line = ((DCacheCount+1) & (DCACHE_SETS-1))));
+    (!doLockDown && (line = (line | DCacheLockDown & (DCACHE_SETS-1))+id)) || (doLockDown && (line = (DCacheLockDown & (DCACHE_SETS-1))+id));
+#endif
+
+    u32* ptr = (u32 *)&DCache[line << DCACHE_LINELENGTH_LOG2];
+
+    DataCycles = 0;
+    #if !DISABLE_CACHEWRITEBACK
+        // Before we fill the cacheline, we need to write back dirty content
+        // Datacycles will be incremented by the required cycles to do so
+        DCacheClearByASetAndWay(line & (DCACHE_SETS-1), line >> DCACHE_SETS_LOG2);
+    #endif
+    //Log(LogLevel::Debug,"DCache miss, load @ %08x\n", tag);
+    for (int i = 0; i < DCACHE_LINELENGTH; i+=sizeof(u32))
+    {
+        if (tag+i < ITCMSize)
+        {
```

Comment:
oh so you can actually put ITCM and DTCM into cache?

Reply:
I am not entirely sure (I think ) I had a block diagram showing the access passing the write buffer before the TCMs but can't seem to find that diagram. If that is the case, the TCM access will go through the cache when enabled.
However my decision to place the read from TCMs here was to keep the access flow within the line fills in sync to the DataReadXX functions.